### PR TITLE
Make Vapor's Client Configurable

### LINKED ISSUE #2269: Support overriding Application.Client
In Vapor 3, it's possible to override the client used by the application, as follows:
`services.register(Client.self) { return MockClient() }`

This is useful for a number of reasons, but is especially handy for testing - the app client can be mocked so requests can be inspected. With the changes to services in Vapor 4, this is no longer possible.

Vapor 4 should support overriding the client in the same way as Vapor 3, possibly by following a similar pattern to that used for settings the view renderer (`app.views.use(.leaf)`).

Possible usage:
`app.client.use(.myFakeClient)`

### PR DESCRIPTION
**Pull request by 0xTim**
This PR makes Vapor's client configurable. This would allow you to switch between Async HTTP Client and URLSession in your apps, and also a real client and a fake client in tests, using something like:

```swift
app.clients.use(.fake)
```

And also demonstrated [in the tests](https://github.com/vapor/vapor/pull/2277/files#diff-c991ce1ca70f01333e0d2a1315004cb2R1332).

This is a wide ranging PR with one breaking changes that can be likely justified as for most people, `req.client` still does what they expect. Anyone using `app.client.configuration` should migrate to `app.clients.configuration` - this allows Vapor's `ClientConfiguration` to be applied to all `Client` types and whoever implements the `Client` protocol can adapt it as they need (though it's source compatible with Async HTTP Client's Configuration to avoid any breaking changes there.

This PR also has the added benefit of abstracting away the Async HTTP Client from the end user. Both `HTTPClient` and `HTTPClient.Configuration` are now mapped to Vapor's own types and these new types are the only ones exposed. Therefore we no longer need to export `AysncHTTPClient` which makes us free to take major updates without having to wait for Vapor 5.

Resolves #2269

### CODE REVIEW

File: Sources/Vapor/Client/Application+Client.swift [hunk 1]
```diff
@@ -42,23 +61,41 @@ extension Application {
...
+            self.storage.makeClient = makeClient
+        }
 
-    public func send(_ request: ClientRequest) -> EventLoopFuture<ClientResponse> {
-        self.http.send(request, eventLoop: .indifferent)
+        public let application: Application
+        
+        private var storage: Storage {
+            guard let storage = self.application.storage[Key.self] else {
+                fatalError("Clients not configured. Configure with app.client.initialize()")
```
Comment (MrLotU) [inline Sources/Vapor/Client/Application+Client.swift]: typo :)
```suggestion
                fatalError("Clients not configured. Configure with app.clients.initialize()")
```

**Comment by MrLotU** [comment]
Maybe rename `Wrapped...` to `Any...`?

**Comment by MrLotU** [comment]
> Anyone using app.client should migrate to app.clients.client - I couldn't find a nice non-breaking way to do this without causing all sorts of confusion.

To avoid breaking you could add an extension to `Application` to add `app.client` back in and link it to `app.clients.client`. Though that might confuse more than it solves

**Comment by 0xTim** [comment]
`AnyHTTPClient` would indicate some sort of type erasure, which I'm not sure is the right impression. We could just call it `HTTPClient` if the compiler can handle that. Or `ASHClient` - basically something to indicate that it's a wrapped version of `AsyncHTTPClient` without directly exposing the type to end users so we don't have to worry about breaking changes in HTTP Client

**Comment by MrLotU** [comment]
I think renaming to `HTTPClient` would be fine (if there's nothing else with that name yet), and I think `ASHClient` could be very confusing.

Also, we're kind of type-erasing right? As in, we erase the fact that this `HTTPClient` is in fact `AsyncHttpClient`

**Comment by 0xTim** [comment]
`HTTPClient` is a type defined in the [Aysnc HTTP Client library](https://github.com/swift-server/async-http-client/blob/master/Sources/AsyncHTTPClient/HTTPClient.swift#L48). So we could in fact just call it `AsyncHTTPClient`

**Comment by MrLotU** [comment]
Good point. Can we remove the export now too? Or do we need it elsewhere

**Comment by 0xTim** [comment]
We need it for the configuration, I'm tempted to try and abstract that way with our own configuration type, but I'll wait to see what others think of that idea first

**Comment by 0xTim** [comment]
I've added back `app.client` and we don't need to export AsyncHTTPClient anymore 🎉 

We implement our own configuration, which is identical to HTTPClients, so there should be no breaking changes but it allows us to adapt that to URLSession for instance

**Comment by 0xTim** [comment]
Closing in favour of #2279 - leaving this for discussion about what Vapor should export or not

**Pull request closed without merging**